Refine handling of local content visibility for posts

pfefferle opened the pull request:
This PR aims to send an `Update` activity when a previously federated post is later set to local, reflecting changes to the audience (`to`, `cc` fields) and `in_reply_to`.

The benefit of this change (if it works as intended) over sending a `Delete` is that the post can be reactivated by simply changing its visibility.

Improves logic for 'activitypub_content_visibility' set to 'local', ensuring posts are only disabled if not already federated. Updates transformer logic to set correct recipients for local posts, prevents in-reply-to for local visibility, and adds/updates related unit tests to cover these scenarios.

<!--- Provide a general summary of your changes in the Title above -->

For reference https://wordpress.org/support/topic/how-to-delete-federated-posts/

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Send Updates when a post was published before and set to `local` afterwards
* Clear `in_reply_to` (not sure if that is needed)
* Only set Post-Author to `to` and `cc` (not if that is needed or if empty arrays would also do the job)

### Other information:

- [X] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Publish a `public` post
* Change visibility to `local`
* Check Outbox if `Update` is properly created

* Write new post and set to `local` before publishing
* Check that no `Update` / `Create` Activities were created

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs crated from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [X] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [X] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Added - for new features
-   [ ] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [X] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Improved handling of posts switched to 'local' visibility so they update correctly and can be re-enabled by changing visibility again.

</details>

pfefferle committed:
Refine handling of local content visibility for posts

Improves logic for 'activitypub_content_visibility' set to 'local', ensuring posts are only disabled if not already federated. Updates transformer logic to set correct recipients for local posts, prevents in-reply-to for local visibility, and adds/updates related unit tests to cover these scenarios.

matticbot committed:
Add changelog

pfefferle commented on includes/transformer/class-base.php:
```diff
@@ -206,6 +206,10 @@ protected function set_audience( $activity_object ) {
 				$activity_object->add_to( $mentions );
 				$activity_object->add_to( $replied_to );
+				break;
+			case ACTIVITYPUB_CONTENT_VISIBILITY_LOCAL:
+				$activity_object->add_to( $actor->get_id() );
+				$activity_object->add_cc( $actor->get_id() );
```
not sure if we have to set the `to` and `cc` fields instead of keeping them empty.

obenland replied:
Yeah me neither. My first instinct would have been to keep them empty. What prompted you to set them?

pfefferle replied:
I am not sure if it will be simply ignored with an empty to and cc, so I made it specific!?

pfefferle commented on includes/transformer/class-post.php:
```diff
@@ -578,6 +578,10 @@ public function generate_reply_link( $block_content, $block ) {
 	 * @return string|array|null The in-reply-to URL of the post.
 	 */
 	protected function get_in_reply_to() {
+		if ( ACTIVITYPUB_CONTENT_VISIBILITY_LOCAL === $this->get_content_visibility() ) {
+			return null;
+		}
```
not sure if we should clear the `in_reply_to` or simply keep it.

obenland commented on includes/functions.php:
```diff
@@ -325,18 +325,29 @@ function is_post_disabled( $post ) {
+		! $disabled &&
+		\in_array( $visibility, array( ACTIVITYPUB_CONTENT_VISIBILITY_LOCAL, ACTIVITYPUB_CONTENT_VISIBILITY_PRIVATE ), true )
+	) {
+		if ( 'federated' === \get_post_meta( $post->ID, 'activitypub_status', true ) ) {
+			// If post was already federated, we need to send an update with the new visibility.
+			$disabled = false;
+
+		} else {
+			$disabled = true;
+		}
+	}
```
```suggestion
	if (
		! $disabled &&
		\in_array( $visibility, array( ACTIVITYPUB_CONTENT_VISIBILITY_LOCAL, ACTIVITYPUB_CONTENT_VISIBILITY_PRIVATE ), true ) &&
		'federated' !== \get_post_meta( $post->ID, 'activitypub_status', true )
	) {
			// If post wasn't federated, we don't need to send an update with the new visibility.
			$disabled = true;
		}
	}
```
Can be simplified since `$disabled` is already false in this conditional.

pfefferle committed:
Update includes/functions.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Fix indentation in is_post_disabled function

Corrected the indentation of a comment and assignment in the is_post_disabled function for better code readability.

pfefferle committed:
Fix handling of local content visibility in transformers

Updates the Base transformer to clear 'to' and 'cc' fields for local content visibility instead of adding the actor's ID. Removes early return in Post transformer for local content visibility in get_in_reply_to().

pfefferle committed:
Update local visibility handling in activity transformer

For activities with local visibility, set the actor's ID as both 'to' and 'cc' recipients instead of empty arrays. This ensures proper addressing for local-only content.

pfefferle commented:
After some feedback from Mastodon contributors: This is actually not possible. We would have to send a `Delete` in this case.

The risk here: A `Delete` is irreversible and a re-publish is not possible then.

pfefferle commented:
@obenland the question is if we want to send a delete!?

obenland commented:
Maybe we could add a notice into the visibility selector in the editor, warning that it's a permanent change?

Other than that, it feels like a specific-enough action to me, to assume that it doesn't happen accidentally. I'd be fine with sending the delete even without a warning.

Jiwoon-Kim commented:
In the case of Misskey, there is an option for delete-and-edit. For WordPress federation, this could be implemented by actually deleting the original post and publishing a duplicate as a new post with a new ID.

Of course, any existing comments or interactions would be lost.

pfefferle committed:
Merge branch 'trunk' into add/send-update-on-visibility-change

Jiwoon-Kim commented:
> For WordPress federation, this could be implemented by actually deleting the original post and publishing a duplicate as a new post with a new ID.

By the way, couldn’t this approach also be applied to actors? I have a hunch that threads.net might be using a similar method to implement federation opt-out.
